parser/src/parser.cpp

```diff
@@ -227,7 +231,17 @@ int main(int argc, char* argv[])
   }
 
   if (!isNewDb)
-    LOG(info) << "Project already exists, incremental parsing in action.";
+  {
+    LOG(info) << "Project already exists, incremental parsing in action"
+              << (vm.count("dry-run") ? " (DRY RUN)" : "") << ".";
+  }
+
+  if (isNewDb && vm.count("dry-run"))
+  {
+    LOG(warning) << "Dry-run can be only used with incremental parsing, "
+                    "no project found, turning --dry-run off.";
+    vm.erase("dry-run");
+  }
```

Comment:
Currently we only detect **modified** and **deleted** files, since we only need to perform any workspace maintenance action regarding these files. **Added files** do not require any maintenance, since they will be correctly handled by the *partial parsing* capability of CodeCompass.

Detecting **modified** and **deleted** files is quite straight-forward, we just need to look up the entries in the `File` table and check the existence and content hash on the disk.
On the other hand, detecting **added** files correctly is non-trivial, since the project source folder on the disk may contain many auxiliary files unrelated to the build progress, which we ignore and do not store in the workspace database. To avoid listing a bunch of auxiliary files as added on each incremental parsing, we would must analyze the build commands to filter out the files which are not part of any build command (; and header files are another case). However, build commands are only parsed and analyzed from the build JSON after the incremental parsing stage, which makes this issue complicated ...

So this feature to list the added files requires quite a good amount of work and we do not need it later during incremental parsing. Therefore on a previous meeting it was agreed (as far as I remember) to not implement it currently just for the listing.

Reply:
Any ideas on a more clever way to deduce this information is very much welcomed though.

Reply:
This is rather a philosophical question. I'd except no database modification in a dry-run.

Reply:
I agree with your assumption, a dry-run shall not perform any modifying actions. My point is that detecting **added** files correctly (without a lot of extra files not part of the compilation process, e.g. image resources) requires a larger amount of work and also some refactoring in the parsing workflow of CodeCompass. This should be done some time, but was currently simply not deemed important enough with the only benefit that we will be able to list the added files on a dry-run.